extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,167 @@
+  }
+
+  executeRealTimeConfig(customMacros) {
+    if (!this.rtcConfig) {
+      return null;
+    }
+    const rtcTimeout = this.rtcConfig.timeoutMillis || DEFAULT_RTC_TIMEOUT;
+    this.inflatePublisherUrls(customMacros);
+    this.inflateVendorUrls();
+    if (this.calloutUrls.length == 0) {
+      return Promise.resolve();
+    }
+    let rtcTime;
+    return Promise.all(this.calloutUrls.map(url => {
+      const rtcStartTime = Date.now();
+      /** ISSUE: For very short timeouts, and/or for slow devices, the calculated time
+       * that RTC took (rtcTime below) can be longer than the timeout without getting timed-out,
+       * due to the time it takes to set up the timeout promise.
+       * EXAMPLE, let rtcTimeout = 100ms.
+       * Time = 0: const rtcStartTime = 0
+       * Time = 10: call Services.timerFor.timeoutPromise
+       * Time = 25: timeout promise starts the xhr callout, timeout promise sees start time as 25
+       * Time = 120: xhr callout returns, timeoutPromise sees total time as 95ms, so no timeout
+       * Time = 125: rtcTime = 125ms, which is greater than the publisher specified timeout.
+       */
+      return Services.timerFor(this.win).timeoutPromise(
+          rtcTimeout,
+          Services.xhrFor(this.win).fetchJson(
+              url, {credentials: 'include'}).then(res => {
+                // Non-200 status codes are forbidden for RTC.
+                // TODO: Add to fetchResponse the ability to
+                // check for redirects as well.
+                if (res.status != 200) {
+                  rtcTime = Date.now() - rtcStartTime;
+                  return {rtcTime, error: 'Non-200 Status', hostname};
+                }
+                return res.text().then(text => {
+                  rtcTime = Date.now() - rtcStartTime;
+                  const hostname = parseUrl(url).hostname;
+                  // An empty text response is fine, just means
+                  // we have nothing to merge.
+                  if (!text) {
+                    return {rtcTime, hostname};
+                  }
+                  const rtcResponse = tryParseJson(text);
+                  return rtcResponse ? {rtcResponse, rtcTime, hostname} :
+                  {rtcTime, hostname, error: 'Unparsable JSON'};
+                });
+              })).catch(err => {
+                const hostname = parseUrl(url).hostname;
+                const rtcTime = Date.now() - rtcStartTime;
+                return {error: err, rtcTime, hostname};
+              });
+    }));
+  }
+
+  /**
+   * Attempts to parse the publisher-defined RTC config off the amp-ad
+   * element, then validates that the rtcConfig exists, and contains
+   * an entry for either vendor URLs, or publisher-defined URLs. If the
+   * config contains an entry for timeoutMillis, validates that it is a
+   * number, or converts to a number if number-like, otherwise overwrites
+   * with the default.
+   * @return {!boolean}
+   */
+  validateRtcConfig() {
```

Comment:
Any reason this can't just be a private function that is called as part of executeRealTimeConfig?  Actually let's do this as part of the constructor so that you store only the timeout and the callouts as an object of keyname (vendor or url domain) to inflated url.  You can also store timeout as a field. E.g:

this.timeoutMillis = 1000;
let configTimeout = rtcConfig['timeoutMillis'];
if (configTimeout != undefined) {
    if (typeof configTimeout != 'number' || configTimeout <= 0 || configTimeout > this.timeoutMillis) {
       user().warn(TAG, `'Invalid RTC timeout: ${configTimeout}, using default ${this.timeoutMillis}`);
    } else {
       this.timeoutMillis = configTimeout;
    }
}